provider/core/src/error.rs

```diff
@@ -255,7 +255,7 @@ impl DataError {
     }
 
     #[inline]
-    pub(crate) fn for_type<T>() -> DataError {
+    pub fn for_type<T>() -> DataError {
```

Comment:
Nit: If you make this public, please:

1. Add docs
2. Consider whether a name like `mismatched_type<T>()` would be better than `for_type<T>()`